portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -39,27 +40,42 @@
   
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+    String defaultCommentLimit = "5";
+    int commentLimit = Integer.parseInt(getParameter(request, "numShown", defaultCommentLimit));
+
+    String defaultPageNumber = "0";
+    int pageNumber = Integer.parseInt(getParameter(request, "pageNumber", defaultPageNumber));
 
+    String defaultSortMethod = "DESCENDING";
+    String sortMethod = getParameter(request, "sortMethod", defaultSortMethod);
+
+    Query query = new Query("Comment");
+    
+    if (sortMethod.equals(defaultSortMethod)) {
+      query.addSort("timestamp", SortDirection.DESCENDING);
+    } else {
+      query.addSort("timestamp", SortDirection.ASCENDING);
+    }
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     PreparedQuery results = datastore.prepare(query);
+    
 
     ArrayList<Comment> comments = new ArrayList<>();
-    String defaultCommentLimit = "5";
-    int commentLimit = Integer.parseInt(getParameter(request, "numShown",
-                                                    "5"));
+
+    int startingIndex = pageNumber * commentLimit;
     int count = 0;
 
     for (Entity entity : results.asIterable()) {
-      long id = entity.getKey().getId();
-      String comment = (String) entity.getProperty("comment");
-      long timestamp = (long) entity.getProperty("timestamp");
-
-      Comment newComment = new Comment(id, comment, timestamp);
-      comments.add(newComment);
+      if (count >= startingIndex) {
```

Comment:
Iterating through all of them is not ideal.  Imagine if the user was on page 200 or something.  Can you turn it into an array?  Or is there an OFFSET keyword for the query?  This is a supported feature in many query languages.

Reply:
Turned query into an array

Reply:
The best way to do this with datastore would be to instantiate with a startCursor and an endCursor. I tried to do this and for various reasons datastore gave me a lot of trouble. I have implemented a middle ground in the meantime: call a query on an upper limit of entries (i.e., endCursor), and then begin the for loop @ start cursor. There is still a space x memory x time hit from populating the query, but it won't loop through unnecessary entities.